<redacted> <redacted>

repo url: https://github.com/<redacted>/assignment-5-judging-flowers-<redacted>

commit id: "10/31 submission"

In [235]:
# Imports and pip installations (if needed)
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import accuracy_score
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier

# Part 1: Load the dataset

In [236]:
# Load the dataset (load remotely or locally)
data = pd.read_csv('iris.csv')

# Output the first 15 rows of the data
print("First 15 rows of data:")
print(data.head(15))

# Display a summary of the table information (number of datapoints, etc.)
print("\nSummary of table information:")
print(data.info())

First 15 rows of data:
    sepal_length  sepal_width  petal_length  petal_width species
0            5.1          3.5           1.4          0.2  setosa
1            4.9          3.0           1.4          0.2  setosa
2            4.7          3.2           1.3          0.2  setosa
3            4.6          3.1           1.5          0.2  setosa
4            5.0          3.6           1.4          0.2  setosa
5            5.4          3.9           1.7          0.4  setosa
6            4.6          3.4           1.4          0.3  setosa
7            5.0          3.4           1.5          0.2  setosa
8            4.4          2.9           1.4          0.2  setosa
9            4.9          3.1           1.5          0.1  setosa
10           5.4          3.7           1.5          0.2  setosa
11           4.8          3.4           1.6          0.2  setosa
12           4.8          3.0           1.4          0.1  setosa
13           4.3          3.0           1.1          0.1  setosa
14

## About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?

The data specifies a number of qualities for many different iris flowers, noting various anatomical iris measurements (sepal length/width and petal length/width), as well as the type of iris species. The features are sepal_length, sepal_width, petal_length, and petal_width. The label is simply the species. The mapping of the label (species) to the actual classes is as follows. Class O corresponds to setosa, class 1 corresponds to versicolor, and class 2 corresponds to virginica.

# Part 2: Split the dataset into train and test

In [237]:
# Take the dataset and split it into our features (X) and label (y)
X = data[['sepal_length', 'sepal_width', 'petal_length', 'petal_width']]
y = data['species']

print("Features (X):")
print(X)
print("\nLabel (y):")
print(y)

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=56)

print("\nX training set:")
print(X_train)
print("\nX testing set:")
print(X_test)
print("\ny training set:")
print(y_train)
print("\ny testing set:")
print(y_test)


Features (X):
     sepal_length  sepal_width  petal_length  petal_width
0             5.1          3.5           1.4          0.2
1             4.9          3.0           1.4          0.2
2             4.7          3.2           1.3          0.2
3             4.6          3.1           1.5          0.2
4             5.0          3.6           1.4          0.2
..            ...          ...           ...          ...
145           6.7          3.0           5.2          2.3
146           6.3          2.5           5.0          1.9
147           6.5          3.0           5.2          2.0
148           6.2          3.4           5.4          2.3
149           5.9          3.0           5.1          1.8

[150 rows x 4 columns]

Label (y):
0         setosa
1         setosa
2         setosa
3         setosa
4         setosa
         ...    
145    virginica
146    virginica
147    virginica
148    virginica
149    virginica
Name: species, Length: 150, dtype: object

X training set:
     sep

# Part 3: Logistic Regression

In [238]:
# i. Use sklearn to train a LogisticRegression model on the training set
label_encoder = LabelEncoder()
y_train_encoded = label_encoder.fit_transform(y_train)
y_test_encoded = label_encoder.transform(y_test)

model = LogisticRegression(max_iter=200)  
model.fit(X_train, y_train_encoded)

# ii. For a sample datapoint, predict the probabilities for each possible class
sample_point = X_test.iloc[[0]] 
probabilities = model.predict_proba(sample_point)

print("Sample datapoint probabilities for each possible class:")
print(probabilities)

# iii. Report on the score for Logistic regression model, what does the score measure?
# SEE MARKDOWN BELOW FOR EXPLANATION OF SCORE
y_pred = model.predict(X_test)
score = accuracy_score(y_test_encoded, y_pred)

print("\nLogistic regression model score:")
print(score)

# iv. Extract the coefficents and intercepts for the boundary line(s)
coefficients = model.coef_
intercepts = model.intercept_

print("\nBoundary line coefficients:")
print(coefficients)
print("\nBoundary line intercepts:")
print(intercepts)

Sample datapoint probabilities for each possible class:
[[7.82443046e-05 1.16268066e-01 8.83653690e-01]]

Logistic regression model score:
0.8666666666666667

Boundary line coefficients:
[[-0.45591719  0.89907511 -2.50161248 -1.01122075]
 [ 0.56670286 -0.26037575 -0.26950061 -0.83260012]
 [-0.11078566 -0.63869936  2.77111309  1.84382087]]

Boundary line intercepts:
[ 10.16453581   1.91975041 -12.08428622]


Boundary line for class 0: $h(x_1, x_2, x_3, x_4) = 10.1645 - 0.4559 \cdot x_1 + 0.8991 \cdot x_2 - 2.5016 \cdot x_3 - 1.0112 \cdot x_4$

Boundary line for class 1: $h(x_1, x_2, x_3, x_4) = 1.9198 + 0.5667 \cdot x_1 - 0.2604 \cdot x_2 - 0.2695 \cdot x_3 - 0.8326 \cdot x_4$

Boundary line for class 2: $h(x_1, x_2, x_3, x_4) = -12.0843 - 0.1108 \cdot x_1 - 0.6387 \cdot x_2 + 2.7711 \cdot x_3 + 1.8438 \cdot x_4$



$x_1 = \text{sepal length}$

$x_2 = \text{sepal width}$

$x_3 = \text{petal length}$

$x_4 = \text{petal width}$


The score for the logistic regression model is 0.867. Since this value is close to 1 (there is a strong correlation between the features and label), it means that the model performs relatively well in terms of making predictions about the species of a given iris based off of its measurements; the model fits the data well. About 86.7% of the predictions were correct.

# Part 4: Support Vector Machine

In [239]:
# i. Use sklearn to train a Support Vector Classifier on the training set
svm_model = SVC(probability=True)
svm_model.fit(X_train, y_train_encoded)

# ii. For a sample datapoint, predict the probabilities for each possible class
svm_sample_point = X_test.iloc[[0]] 
svm_probabilities = svm_model.predict_proba(svm_sample_point)

print("Sample datapoint probabilities for each possible class:")
print(svm_probabilities)

# iii. Report on the score for the SVM, what does the score measure?
#SEE MARKDOWN BELOW FOR EXPLANATION OF SCORE
svm_y_pred = svm_model.predict(X_test)
svm_score = accuracy_score(y_test_encoded, svm_y_pred)

print("\nSVM score:")
print(svm_score)

Sample datapoint probabilities for each possible class:
[[0.00833175 0.02405285 0.9676154 ]]

SVM score:
0.9333333333333333


The score for the SVM is 0.933. Since this value is close to 1 (there is a strong correlation between the features and label), it means that the model performs very well in terms of making predictions about the species of a given iris based off of its measurements; the model fits the data well. About 93.3% of the predictions were correct.

# Part 5: Neural Network

In [240]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
mlp_model = MLPClassifier(max_iter=1000, random_state=56) 
mlp_model.fit(X_train, y_train_encoded)

# ii. For a sample datapoint, predict the probabilities for each possible class
mlp_sample_point = X_test.iloc[[0]] 
mlp_probabilities = mlp_model.predict_proba(mlp_sample_point)

print("Sample datapoint probabilities for each possible class:")
print(mlp_probabilities)

# iii. Report on the score for the Neural Network, what does the score measure?
#SEE MARKDOWN BELOW FOR EXPLANATION OF SCORE
mlp_y_pred = mlp_model.predict(X_test)
mlp_score = accuracy_score(y_test_encoded, mlp_y_pred)

print("\nNeural Network score:")
print(mlp_score)

# iv: Experiment with different options for the neural network, report on your best configuration
#SEE MARKDOWN BELOW FOR REPORT ON BEST CONFIGURATION    
best_score = 0
best_config = None

print("\nTesting configurations...")
for hidden_layers in [(10,), (20,), (10, 10), (20, 20), (10, 10, 10)]:
    for activation in ['relu', 'tanh']:
        temp_mlp = MLPClassifier(hidden_layer_sizes=hidden_layers, activation=activation, max_iter=2000, random_state=56, early_stopping=True, validation_fraction=0.1, n_iter_no_change=10)
        temp_mlp.fit(X_train, y_train_encoded)
        
        temp_y_pred = temp_mlp.predict(X_test)
        temp_score = accuracy_score(y_test_encoded, temp_y_pred)
        
        print(f"Configuration: hidden_layers={hidden_layers}, activation='{activation}' - Accuracy: {temp_score}")
        
        if temp_score > best_score:
            best_score = temp_score
            best_config = (hidden_layers, activation)

print(f"\nBest configuration: Hidden layers={best_config[0]}, Activation='{best_config[1]}' - Score: {best_score}")

Sample datapoint probabilities for each possible class:
[[1.88803248e-05 5.50492566e-02 9.44931863e-01]]

Neural Network score:
0.9333333333333333

Testing configurations...
Configuration: hidden_layers=(10,), activation='relu' - Accuracy: 0.3333333333333333
Configuration: hidden_layers=(10,), activation='tanh' - Accuracy: 0.3333333333333333
Configuration: hidden_layers=(20,), activation='relu' - Accuracy: 0.4
Configuration: hidden_layers=(20,), activation='tanh' - Accuracy: 0.26666666666666666
Configuration: hidden_layers=(10, 10), activation='relu' - Accuracy: 0.4
Configuration: hidden_layers=(10, 10), activation='tanh' - Accuracy: 0.4
Configuration: hidden_layers=(20, 20), activation='relu' - Accuracy: 0.26666666666666666
Configuration: hidden_layers=(20, 20), activation='tanh' - Accuracy: 0.8
Configuration: hidden_layers=(10, 10, 10), activation='relu' - Accuracy: 0.26666666666666666
Configuration: hidden_layers=(10, 10, 10), activation='tanh' - Accuracy: 0.9333333333333333

Best c

The score for the first neural network is 0.933. Since this value is close 1 (there is a strong correlation between the features and label), it means that the model performs well in terms of making predictions about the species of a given iris based off of its measurements; the model fits the data very well. About 93.3% of the predictions were correct.

In order to experiment with different options for the neural network, a for loop was utilized to iterate through various configurations with different activation functions and sizes of hidden layers. Each of these configurations was evaluated based on the model score. Some models scored as low as 0.267, 0.333, and 0.4, while some had higher scores of 0.8 and 0.933. Ultimately, the best configuration of the options explored utilized the tanh activation function and had 3 hidden layers with 10 nodes each. This configuration had a model score of 0.933. Since this value is close to 1 (there is a strong correlation between the features and label), it means that this configuration performs well in terms of making predictions about the species of a given iris based off of its measurements; the model fits the data well. About 93.3% of the predictions were correct.

# Part 6: K-Nearest Neighbors

In [241]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier
knn_model = KNeighborsClassifier(n_neighbors=60) 
knn_model.fit(X_train, y_train_encoded) 

# ii. For a sample datapoint, predict the probabilities for each possible class
knn_sample_point = X_test.iloc[[0]]  # Use double brackets to retain DataFrame structure
knn_probabilities = knn_model.predict_proba(knn_sample_point)

print("Sample datapoint probabilities for each possible class:")
print(knn_probabilities)

# iii. Report on the score for kNN, what does the score measure?
#SEE MARKDOWN BELOW FOR EXPLANATION OF SCORE
knn_y_pred = knn_model.predict(X_test)
knn_score = accuracy_score(y_test_encoded, knn_y_pred)

print("\nkNN score:")
print(knn_score)

Sample datapoint probabilities for each possible class:
[[0.         0.38333333 0.61666667]]

kNN score:
0.8


The score for the kNN is 0.8. Since this value is relatively close to 1 (there is a strong correlation between the features and label), it means that the model performs relatively well in terms of making predictions about the species of a given iris based off of its measurements; the model fits the data well. About 80,0% of the predictions were correct.

# Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

The score of each model is described in the markdown below it. Ultimately, the support vector machine and neural network models performed the best and similarly at that, each with a high score of 0.933. As stated, it means that these model perform very well in terms of making predictions about the species of a given iris based off of its measurements. The models fit the data well and about 93.3% of the predictions were correct for each of these models. On the other hand the logistic regression model had a slightly lower score of 0.867, meaning this model performs relatively well in terms of making predictions about the species of a given iris based off of its measurements, though not as accurately as the previous described models. The logistic regression model still seems to fit the data relatively well and about 86.7% of the predictions were correct. The logistic regression model likely underperformed the other models due to the fact that logistic regression is linear, making it more challenging to pick up on more subtle complexities in the data. It is likely the case that the relationships between features are complex and far from linear, making it harder for the logistic regression model to keep up with the other models. The k-nearest neighbors model also had a slightly lower score (lower than logistic regression as well) of 0.800. This model also performs relatively well in terms of making predictions about the species of a given iris based off of its measurements, though not as accurately as the previous described models. This model still seems to fit the data relatively well and about 80.0% of the predictions were correct. The k-nearest neighbors model likely underperformed the other models due to a suboptimal choice of k. This hyperparameter can potentially be adjusted and tuned to achieve better performance. I was surprised that this model performed the worst out of all of the models even though it is a nonparametric model because this dataset appears to be relatively simple with few features.